ReactiveCocoa/Swift/Signal.swift

```diff
@@ -410,16 +410,127 @@ public func takeUntil<T, E>(trigger: Signal<(), NoError>)(signal: Signal<T, E>)
+public func reduce<T, U, E>(initial: U, combine: (U, T) -> U)(signal: Signal<T, E>) -> Signal<U, E> {
+	return signal
+		|> scan(initial, combine)
+		|> takeLast(1)
+}
+
+/// Aggregates `signal`'s values into a single combined value. When `signal` emits
+/// its first value, `combine` is invoked with `initial` as the first argument and
+/// that emitted value as the second argument. The result is emitted from the
+/// signal returned from `reduce`. That result is then passed to `combine` as the
+/// first argument when the next value is emitted, and so on.
+public func scan<T, U, E>(initial: U, combine: (U, T) -> U)(signal: Signal<T, E>) -> Signal<U, E> {
+	return Signal { observer in
+		let accumulatorState = Atomic(initial)
+		return signal.observe(next: { value in
+			accumulatorState.modify { accumulator in
+				let newAccumulatorValue = combine(accumulator, value)
+				sendNext(observer, newAccumulatorValue)
+				return newAccumulatorValue
+			}
+			return
+		}, error: { error in
+			sendError(observer, error)
+		}, completed: {
+			sendCompleted(observer)
+		})
+	}
+}
+
+/// Forwards only those values from `signal` which are not duplicates of the
+/// immedately preceding value. The first value is always forwarded.
+public func skipRepeats<T: Equatable, E>(signal: Signal<T, E>) -> Signal<T, E> {
+	return signal |> skipRepeats { $0 == $1 }
+}
+
+/// Forwards only those values from `signal` which do not pass `isRepeat` with
+/// respect to the previous value. The first value is always forwarded.
+public func skipRepeats<T, E>(isRepeat: (T, T) -> Bool)(signal: Signal<T, E>) -> Signal<T, E> {
+	return signal
+		|> map { Optional($0) }
+		|> combinePrevious(nil)
+		|> filter {
+			switch $0 {
+			case let (.Some(a), .Some(b)) where isRepeat(a, b):
+				return false
+			default:
+				return true
+			}
+		}
+		|> map { $0.1! }
+}
+
+/// Does not forward any values from `signal` until `predicate` returns false,
+/// at which point the returned signal behaves exactly like `signal`.
+public func skipWhile<T, E>(predicate: T -> Bool)(signal: Signal<T, E>) -> Signal<T, E> {
+	return Signal { observer in
+		let isSkipping = Atomic(true)
+		return signal.observe(next: { value in
+			isSkipping.modify { wasSkipping in
+				var shouldSkip = wasSkipping && predicate(value)
+				if !shouldSkip {
+					sendNext(observer, value)
+				}
+				return shouldSkip
+			}
+			return
+		}, error: { error in
+			sendError(observer, error)
+		}, completed: {
+			sendCompleted(observer)
+		})
+	}
+}
+
+/// Waits until `signal` completes and then forwards the final `count` values
+/// on the returned signal.
+public func takeLast<T,E>(count: Int)(signal: Signal<T,E>) -> Signal<T,E> {
+	return Signal { observer in
+		let buffer = Atomic<[T]>({
+			var bufferArray = [T]()
+			bufferArray.reserveCapacity(count)
+			return bufferArray
+		}())
+		
+		return signal.observe(next: { value in
+			buffer.modify { oldBuffer in
+				var newBuffer = oldBuffer
```

Comment:
Yeah. It will only be one real copy because Swift is smart about avoiding spurious copies.